### PR TITLE
[Jon Chua] Duke Increments

### PR DESCRIPTION
**Pull request by jon-chua**
- Implemented Duke Levels 1 to 6

### CODE REVIEW

File: src/main/java/duke/command/AddDeadlineCommand.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import duke.main.TaskList;
+import duke.main.Ui;
+import duke.task.Deadline;
+
+import java.util.Date;
+
+public class AddDeadlineCommand extends Command {
+    private String description;
+    private String deadlineBy;
+    private Date deadlineDate;
```
Comment (LeowWB) [inline src/main/java/duke/command/AddDeadlineCommand.java:14]: Between `deadlineDate` and `deadlineBy`, it seems like one of them will always be null. Maybe you should consider just using a single variable?

File: src/main/java/duke/command/AddEventCommand.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import duke.main.TaskList;
+import duke.main.Ui;
+import duke.task.Event;
+
+import java.util.Date;
+
+public class AddEventCommand extends Command {
+    private String description;
+    private String eventAt;
+    private Date eventDate;
```
Comment (LeowWB) [inline src/main/java/duke/command/AddEventCommand.java:14]: Similar to `AddDeadlineCommand`.

File: src/main/java/duke/command/DeleteCommand.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import duke.main.Storage;
+import duke.main.TaskList;
+import duke.main.Ui;
+import duke.task.Task;
+
+public class DeleteCommand extends Command {
+    private int zeroBasedIndex;
+
+    public DeleteCommand(int oneBasedIndex) {
+        this.zeroBasedIndex = oneBasedIndex - 1;
```
Comment (LeowWB) [inline src/main/java/duke/command/DeleteCommand.java:13]: Have to say that I really like this method of naming your indices. Good job 👍👍👍

File: src/main/java/duke/main/Parser.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+import duke.command.ListCommand;
+import duke.exception.DukeException;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+class Parser {
+    static Command parse(String fullCommand) throws DukeException {
+        if (isExitCommand(fullCommand)) {
```
Comment (LeowWB) [inline src/main/java/duke/main/Parser.java:21]: In my opinion, this could be better expressed as a `switch` statement. Maybe you could add some `String` or `enum` identifier to each of your command classes to facilitate this.

**Comment by LeowWB** [comment]
Overall thoughts:

Really like the extent of your abstraction, especially for your command classes (and how each of them had an `execute()` function). I'd say your testing is also quite on-point; you have quite a good variety of unit test cases that should prevent your individual components for breaking. One thing you may want to try to do in general is to write more documentation, as I noticed quite a few `public` methods missing JavaDoc comments.

File: src/main/java/duke/main/Parser.java [hunk 5]
```diff
@@ -67,7 +67,7 @@ private static Date parseDate(String date) throws DukeException {
         }
     }
 
-    private static boolean isDate(String input) {
+    static boolean isDate(String input) {
```
Comment (caesarpjz) [inline src/main/java/duke/main/Parser.java:75]: Great code all around! :) Maybe you could consider the edge cases for date, for example, if the user enter something like 45/52/3012 1800 by either deadline or event.

File: src/test/java/duke/main/ParserTest.java [hunk 6]
```diff
@@ -18,6 +19,7 @@
     void parserTest() throws DukeException  {
```
Comment (caesarpjz) [inline src/test/java/duke/main/ParserTest.java:19]: Wow! Didn't know you could check for class types. Good job!

**Comment by caesarpjz** [comment]
In general, I think your code is very well documented and neatly arranged. Your choice of method names are very clear and is very easily understood by anyone reading it. It was quite impressive that you had many smaller methods for several commands. Thanks for the good read!
